cmd/acr/purge.go

```diff
@@ -568,38 +586,48 @@ func dryRunPurge(ctx context.Context, acrClient api.AcrCLIClientInterface, login
 			}
 			return -1, -1, err
 		}
-		// This will act as a set if a key is present then it should not be deleted because it is referenced by a multiarch manifest
+		// This will act as a set if a key is present then it should not be deleted
+		// because it is referenced by a multiarch manifest or a manifest has subject attached
 		// that will not be deleted
-		doNotDelete := map[string]bool{}
+		doNotDelete := make(map[string]struct{})
 		candidatesToDelete := []acr.ManifestAttributesBase{}
 		// Iterate over all manifests to discover multiarchitecture manifests
 		for resultManifests != nil && resultManifests.ManifestsAttributes != nil {
 			manifests := *resultManifests.ManifestsAttributes
 			for _, manifest := range manifests {
-				// If the manifest is manifest list and would not get deleted then mark it's dependant manifests as not deletable.
-				if (*manifest.MediaType == manifestListContentType || *manifest.MediaType == manifestIndexContentType) && (*countMap)[*manifest.Digest] != deletedTags[*manifest.Digest] {
-					var manifestListBytes []byte
-					manifestListBytes, err = acrClient.GetManifest(ctx, repoName, *manifest.Digest)
-					if err != nil {
-						return -1, -1, err
-					}
-					var manifestList multiArchManifest
-					err = json.Unmarshal(manifestListBytes, &manifestList)
-					if err != nil {
-						return -1, -1, err
-					}
-					for _, dependentDigest := range manifestList.Manifests {
-						doNotDelete[dependentDigest.Digest] = true
+				if (*countMap)[*manifest.Digest] != deletedTags[*manifest.Digest] {
+					if _, ok := mediaTypes[*manifest.MediaType]; ok {
+						manifestBytes, err := acrClient.GetManifest(ctx, repoName, *manifest.Digest)
+						if err != nil {
+							return -1, -1, err
+						}
+						var customManifest customManifest
+						err = json.Unmarshal(manifestBytes, &customManifest)
+						if err != nil {
+							return -1, -1, err
+						}
+						if customManifest.Subject != nil {
+							// if manifest has subject, manifest is marked to not be deleted
+							doNotDelete[*manifest.Digest] = struct{}{}
+						} else if manifest.Tags == nil {
+							// If the manifest has no subject or tag, it is a candidate for deletion
+							candidatesToDelete = append(candidatesToDelete, manifest)
+						}
+						if len(customManifest.Manifests) > 0 {
```

Comment:
This is happening irrespective of if the manifest has been marked for deletion in the previous step

Reply:
yes. similar to comment above that some manifests could be double added to candidatesToDelete. addressed in the filterManifests method.